### PR TITLE
Impl Eq and Hash for all evaluable

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

This PR implements Eq and Hash for all evaluable types, so that they can be used in HashMaps.

About the change in JSON:

- According to https://github.com/pingcap/tidb/issues/10461 MySQL does not do epsilon aware comparing.
- According to [TiDB's implementation](https://github.com/pingcap/tidb/blob/545afc5a3a0fc806b02324c39d6dd01387a04b80/types/json/binary_functions.go#L638) and my own experiment, we should not cast data to f64 any more when data is not numeric. We should only do precedence compare.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Existing unit tests (some incorrect cases are dropped).

### CODE REVIEW

File: src/coprocessor/codec/mysql/json/comparison.rs [hunk 1]
```diff
@@ -63,17 +63,14 @@ impl PartialEq for Json {
...
 
 impl PartialOrd for Json {
+    // FIXME: The following logic is not exactly the same as MySQL.
     fn partial_cmp(&self, right: &Json) -> Option<Ordering> {
         let precedence_diff = self.get_precedence() - right.get_precedence();
         if precedence_diff == 0 {
             if self.get_precedence() == PRECEDENCE_NUMBER {
                 let left_data = self.as_f64().unwrap();
                 let right_data = right.as_f64().unwrap();
-                if (left_data - right_data).abs() < f64::EPSILON {
```
Comment (AndreMouche) [inline src/coprocessor/codec/mysql/json/comparison.rs]: why remove it?

**Comment by AndreMouche** [comment]
for the change of `json` @zz-jason what do you think

**Comment by breeswish** [comment]
Close since fuzz test shown that Decimal's Hash implementation is not correct and we are not relying on Hash for Decimal / DateTime / Json any more in https://github.com/tikv/tikv/pull/4752

**Pull request closed without merging**